modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java

```diff
@@ -1510,18 +1510,28 @@ public DDMStructure updateStructure(
 	@Override
 	public DDMStructure updateStructure(
 			long userId, long structureId, long parentStructureId,
-			Map<Locale, String> nameMap, Map<Locale, String> descriptionMap,
-			String definition, ServiceContext serviceContext)
+			String structureKey, Map<Locale, String> nameMap,
+			Map<Locale, String> descriptionMap, String definition,
+			ServiceContext serviceContext)
 		throws PortalException {
 
 		DDMStructure structure = ddmStructurePersistence.findByPrimaryKey(
 			structureId);
 
 		User user = _userLocalService.getUser(userId);
 
+		_validate(
+			structure.getGroupId(), structure.getParentStructureId(),
+			structure.getClassNameId(), structureKey, nameMap,
+			structure.getDDMForm());
+
```

Comment:
@kristofloneck what happens if we pass the current value of `structure.getStructureKey()` to `structureKey`?

I suspect that in this case, [ddmStructurePersistence.fetchByG_C_S](https://github.com/liferay/liferay-portal/blob/570b5a7c2db64db5c62b4494c3cc29d12cbc17df/modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/service/impl/DDMStructureLocalServiceImpl.java#L2072-L2075) will return a non-null structure and the `structureDuplicateStructureKeyException` will be thrown when it shouldn't, because the returned structure is the same one we are updating.

We could fix this by doing:

```
DDMStructure structure = ddmStructurePersistence.fetchByG_C_S(
	groupId, classNameId, structureKey);

if ((structure != null) && (structure.getStructureId() != currentStructureId)) {
	StructureDuplicateStructureKeyException
		structureDuplicateStructureKeyException =
			new StructureDuplicateStructureKeyException();

	structureDuplicateStructureKeyException.setStructureKey(
		structure.getStructureKey());

	throw structureDuplicateStructureKeyException;
}
```